File: src/indexer.ts
import fg from "fast-glob";
import { readFile, stat } from "node:fs/promises";
import { basename, dirname, extname } from "node:path";
import type { IndexedFile, CodeSymbol, SymbolType } from "./types.js";

export class CodeIndexer {
	private cache = new Map<string, IndexedFile>();

	async indexDirectory(
		directory: string,
		patterns: string[] = ["**/*.{ts,js,tsx,jsx,py,rs,go,java,cpp,c,h}"],
	): Promise<void> {
		const files = await fg(patterns, {
			cwd: directory,
			absolute: true,
			ignore: ["node_modules/**", "dist/**", "build/**", ".git/**"],
		});

		await Promise.all(files.map((file) => this.indexFile(file)));
	}

	async indexFile(filePath: string): Promise<void> {
		try {
			const stats = await stat(filePath);
			const existing = this.cache.get(filePath);

			if (existing && existing.lastModified >= stats.mtime.getTime()) {
				return;
			}

			const content = await readFile(filePath, "utf-8");
			const symbols = await this.extractSymbols(filePath, content);

			// Add filename and dirname as symbols
			const fileSymbols: CodeSymbol[] = [
				{
					name: basename(filePath),
					type: "filename",
					file: filePath,
					line: 1,
					column: 1,
				},
				{
					name: basename(dirname(filePath)),
					type: "dirname",
					file: filePath,
					line: 1,
					column: 1,
				},
				...symbols,
			];

			this.cache.set(filePath, {
				path: filePath,
				symbols: fileSymbols,
				lastModified: stats.mtime.getTime(),
			});
		} catch (error) {
			console.warn(`Failed to index ${filePath}:`, error);
		}
	}

	private async extractSymbols(
		filePath: string,
		content: string,
	): Promise<CodeSymbol[]> {
		const ext = extname(filePath).toLowerCase();
		const symbols: CodeSymbol[] = [];

		switch (ext) {
			case ".ts":
			case ".tsx":
			case ".js":
			case ".jsx":
				symbols.push(...this.extractJavaScriptSymbols(filePath, content));
				break;
			case ".py":
				symbols.push(...this.extractPythonSymbols(filePath, content));
				break;
			case ".rs":
				symbols.push(...this.extractRustSymbols(filePath, content));
				break;
			case ".go":
				symbols.push(...this.extractGoSymbols(filePath, content));
				break;
			default:
				symbols.push(...this.extractGenericSymbols(filePath, content));
		}

		return symbols;
	}

	private extractJavaScriptSymbols(
		filePath: string,
		content: string,
	): CodeSymbol[] {
		const symbols: CodeSymbol[] = [];
		const lines = content.split("\n");

		for (let i = 0; i < lines.length; i++) {
			const line = lines[i];
			if (!line) continue;
			const lineNumber = i + 1;

			// Functions
			const functionMatch = line.match(
				/(?:function\s+|const\s+|let\s+|var\s+)(\w+)\s*(?:=\s*(?:async\s+)?(?:function|\()|(?:\(.*?\)\s*=>))/,
			);
			if (functionMatch && functionMatch[1]) {
				symbols.push({
					name: functionMatch[1],
					type: "function",
					file: filePath,
					line: lineNumber,
					column: (functionMatch.index ?? 0) + 1,
					context: line.trim(),
				});
			}

			// Classes
			const classMatch = line.match(/class\s+(\w+)/);
			if (classMatch && classMatch[1]) {
				symbols.push({
					name: classMatch[1],
					type: "class",
					file: filePath,
					line: lineNumber,
					column: (classMatch.index ?? 0) + 1,
					context: line.trim(),
				});
			}

			// Interfaces
			const interfaceMatch = line.match(/interface\s+(\w+)/);
			if (interfaceMatch && interfaceMatch[1]) {
				symbols.push({
					name: interfaceMatch[1],
					type: "interface",
					file: filePath,
					line: lineNumber,
					column: (interfaceMatch.index ?? 0) + 1,
					context: line.trim(),
				});
			}

			// Type aliases
			const typeMatch = line.match(/type\s+(\w+)/);
			if (typeMatch && typeMatch[1]) {
				symbols.push({
					name: typeMatch[1],
					type: "type",
					file: filePath,
					line: lineNumber,
					column: (typeMatch.index ?? 0) + 1,
					context: line.trim(),
				});
			}

			// Enums
			const enumMatch = line.match(/enum\s+(\w+)/);
			if (enumMatch && enumMatch[1]) {
				symbols.push({
					name: enumMatch[1],
					type: "enum",
					file: filePath,
					line: lineNumber,
					column: (enumMatch.index ?? 0) + 1,
					context: line.trim(),
				});
			}

			// Variables/Constants
			const varMatch = line.match(/(?:const|let|var)\s+(\w+)/);
			if (varMatch && varMatch[1] && !functionMatch) {
				symbols.push({
					name: varMatch[1],
					type: line.includes("const") ? "constant" : "variable",
					file: filePath,
					line: lineNumber,
					column: (varMatch.index ?? 0) + 1,
					context: line.trim(),
				});
			}
		}

		return symbols;
	}

	private extractPythonSymbols(
		filePath: string,
		content: string,
	): CodeSymbol[] {
		const symbols: CodeSymbol[] = [];
		const lines = content.split("\n");

		for (let i = 0; i < lines.length; i++) {
			const line = lines[i];
			if (!line) continue;
			const lineNumber = i + 1;

			// Functions
			const functionMatch = line.match(/def\s+(\w+)/);
			if (functionMatch && functionMatch[1]) {
				symbols.push({
					name: functionMatch[1],
					type: "function",
					file: filePath,
					line: lineNumber,
					column: (functionMatch.index ?? 0) + 1,
					context: line.trim(),
				});
			}

			// Classes
			const classMatch = line.match(/class\s+(\w+)/);
			if (classMatch && classMatch[1]) {
				symbols.push({
					name: classMatch[1],
					type: "class",
					file: filePath,
					line: lineNumber,
					column: (classMatch.index ?? 0) + 1,
					context: line.trim(),
				});
			}

			// Variables
			const varMatch = line.match(/^(\w+)\s*=/);
			if (varMatch && varMatch[1] && !functionMatch && !classMatch) {
				symbols.push({
					name: varMatch[1],
					type: line.match(/^[A-Z_]+\s*=/) ? "constant" : "variable",
					file: filePath,
					line: lineNumber,
					column: (varMatch.index ?? 0) + 1,
					context: line.trim(),
				});
			}
		}

		return symbols;
	}

	private extractRustSymbols(filePath: string, content: string): CodeSymbol[] {
		const symbols: CodeSymbol[] = [];
		const lines = content.split("\n");

		for (let i = 0; i < lines.length; i++) {
			const line = lines[i];
			if (!line) continue;
			const lineNumber = i + 1;

			// Functions
			const functionMatch = line.match(/fn\s+(\w+)/);
			if (functionMatch && functionMatch[1]) {
				symbols.push({
					name: functionMatch[1],
					type: "function",
					file: filePath,
					line: lineNumber,
					column: (functionMatch.index ?? 0) + 1,
					context: line.trim(),
				});
			}

			// Structs
			const structMatch = line.match(/struct\s+(\w+)/);
			if (structMatch && structMatch[1]) {
				symbols.push({
					name: structMatch[1],
					type: "class",
					file: filePath,
					line: lineNumber,
					column: (structMatch.index ?? 0) + 1,
					context: line.trim(),
				});
			}

			// Enums
			const enumMatch = line.match(/enum\s+(\w+)/);
			if (enumMatch && enumMatch[1]) {
				symbols.push({
					name: enumMatch[1],
					type: "enum",
					file: filePath,
					line: lineNumber,
					column: (enumMatch.index ?? 0) + 1,
					context: line.trim(),
				});
			}
		}

		return symbols;
	}

	private extractGoSymbols(filePath: string, content: string): CodeSymbol[] {
		const symbols: CodeSymbol[] = [];
		const lines = content.split("\n");

		for (let i = 0; i < lines.length; i++) {
			const line = lines[i];
			if (!line) continue;
			const lineNumber = i + 1;

			// Functions
			const functionMatch = line.match(/func\s+(\w+)/);
			if (functionMatch && functionMatch[1]) {
				symbols.push({
					name: functionMatch[1],
					type: "function",
					file: filePath,
					line: lineNumber,
					column: (functionMatch.index ?? 0) + 1,
					context: line.trim(),
				});
			}

			// Types
			const typeMatch = line.match(/type\s+(\w+)/);
			if (typeMatch && typeMatch[1]) {
				symbols.push({
					name: typeMatch[1],
					type: "type",
					file: filePath,
					line: lineNumber,
					column: (typeMatch.index ?? 0) + 1,
					context: line.trim(),
				});
			}

			// Variables
			const varMatch = line.match(/var\s+(\w+)/);
			if (varMatch && varMatch[1]) {
				symbols.push({
					name: varMatch[1],
					type: "variable",
					file: filePath,
					line: lineNumber,
					column: (varMatch.index ?? 0) + 1,
					context: line.trim(),
				});
			}
		}

		return symbols;
	}

	private extractGenericSymbols(
		filePath: string,
		content: string,
	): CodeSymbol[] {
		const symbols: CodeSymbol[] = [];
		const lines = content.split("\n");

		for (let i = 0; i < lines.length; i++) {
			const line = lines[i];
			if (!line) continue;
			const lineNumber = i + 1;

			// Generic identifier extraction (simple word boundaries)
			const identifiers = line.match(/\b[a-zA-Z_][a-zA-Z0-9_]*\b/g);
			if (identifiers) {
				for (const identifier of identifiers) {
					if (identifier.length > 2) {
						// Skip very short identifiers
						symbols.push({
							name: identifier,
							type: "variable",
							file: filePath,
							line: lineNumber,
							column: line.indexOf(identifier) + 1,
							context: line.trim(),
						});
					}
				}
			}
		}

		return symbols;
	}

	getAllSymbols(): CodeSymbol[] {
		const allSymbols: CodeSymbol[] = [];
		for (const indexedFile of this.cache.values()) {
			allSymbols.push(...indexedFile.symbols);
		}
		return allSymbols;
	}

	getSymbolsByFile(filePath: string): CodeSymbol[] {
		return this.cache.get(filePath)?.symbols ?? [];
	}

	clearCache(): void {
		this.cache.clear();
	}
}

File: src/indexer.rs
use crate::types::*;
use crate::parsers::SymbolExtractor;
use crate::filters::{FileFilter, GitignoreFilter};
use std::path::{Path, PathBuf};
use std::collections::HashMap;
use std::fs;
use anyhow::{Result, anyhow};
use rayon::prelude::*;

pub struct TreeSitterIndexer {
    symbols_cache: HashMap<PathBuf, Vec<CodeSymbol>>,
    initialized: bool,
    verbose: bool,
    respect_gitignore: bool,
    symbol_extractor: SymbolExtractor,
    file_filter: FileFilter,
    gitignore_filter: GitignoreFilter,
}

impl TreeSitterIndexer {
    pub fn new() -> Self {
        let verbose = false;
        let respect_gitignore = true;
        
        Self {
            symbols_cache: HashMap::new(),
            initialized: false,
            verbose,
            respect_gitignore,
            symbol_extractor: SymbolExtractor::new(verbose),
            file_filter: FileFilter::new(verbose),
            gitignore_filter: GitignoreFilter::new(respect_gitignore, verbose),
        }
    }
    
    pub fn with_verbose(verbose: bool) -> Self {
        let respect_gitignore = true;
        
        Self {
            symbols_cache: HashMap::new(),
            initialized: false,
            verbose,
            respect_gitignore,
            symbol_extractor: SymbolExtractor::new(verbose),
            file_filter: FileFilter::new(verbose),
            gitignore_filter: GitignoreFilter::new(respect_gitignore, verbose),
        }
    }

    pub fn with_options(verbose: bool, respect_gitignore: bool) -> Self {
        Self {
            symbols_cache: HashMap::new(),
            initialized: false,
            verbose,
            respect_gitignore,
            symbol_extractor: SymbolExtractor::new(verbose),
            file_filter: FileFilter::new(verbose),
            gitignore_filter: GitignoreFilter::new(respect_gitignore, verbose),
        }
    }

    pub async fn initialize(&mut self) -> Result<()> {
        self.initialized = true;
        Ok(())
    }
    
    // Synchronous version for parallel processing
    pub fn initialize_sync(&mut self) -> Result<()> {
        self.initialized = true;
        Ok(())
    }

    pub async fn index_file(&mut self, file_path: &Path) -> Result<()> {
        if !self.initialized {
            return Err(anyhow!("Indexer not initialized"));
        }

        let symbols = self.create_file_symbols(file_path)?;
        self.symbols_cache.insert(file_path.to_path_buf(), symbols);
        Ok(())
    }
    
    // Synchronous version for parallel processing that returns symbols
    pub fn index_file_sync(&mut self, file_path: &Path) -> Result<Vec<CodeSymbol>> {
        if !self.initialized {
            return Err(anyhow!("Indexer not initialized"));
        }

        self.create_file_symbols(file_path)
    }
    
    // Public synchronous method for extracting symbols without caching
    pub fn extract_symbols_sync(&self, file_path: &Path, verbose: bool) -> Result<Vec<CodeSymbol>> {
        // Use create_file_symbols to get complete symbols including filename/dirname
        let symbols = self.create_file_symbols(file_path)?;

        if verbose && !symbols.is_empty() {
            eprintln!("Extracted {} symbols from {}", symbols.len(), file_path.display());
        }

        Ok(symbols)
    }

    /// Create file and directory symbols for a given path
    fn create_file_symbols(&self, file_path: &Path) -> Result<Vec<CodeSymbol>> {
        // Handle non-existent files gracefully
        if !file_path.exists() {
            return Ok(vec![]);
        }

        let mut symbols = Vec::new();

        // Add filename and dirname symbols
        if let Some(filename) = file_path.file_name() {
            symbols.push(CodeSymbol {
                name: filename.to_string_lossy().to_string(),
                symbol_type: SymbolType::Filename,
                file: file_path.to_path_buf(),
                line: 1,
                column: 1,
                context: None,
            });
        }

        if let Some(parent) = file_path.parent() {
            if let Some(dirname) = parent.file_name() {
                symbols.push(CodeSymbol {
                    name: dirname.to_string_lossy().to_string(),
                    symbol_type: SymbolType::Dirname,
                    file: file_path.to_path_buf(),
                    line: 1,
                    column: 1,
                    context: None,
                });
            }
        }

        // Get file extension for parser selection
        let extension = file_path.extension()
            .and_then(|ext| ext.to_str())
            .unwrap_or("");

        // For supported extensions, parse the file using Tree-sitter
        match extension {
            // Web languages
            "ts" | "tsx" | "js" | "jsx" | "py" | "php" | "rb" | "ruby" => {
                if let Ok(source_code) = fs::read_to_string(file_path) {
                    if let Ok(extracted) = self.symbol_extractor.extract_symbols(&source_code, file_path) {
                        symbols.extend(extracted);
                    }
                }
            }
            // Systems languages
            "go" | "rs" | "java" | "c" | "cpp" | "cc" | "cxx" | "h" | "hpp" => {
                if let Ok(source_code) = fs::read_to_string(file_path) {
                    if let Ok(extracted) = self.symbol_extractor.extract_symbols(&source_code, file_path) {
                        symbols.extend(extracted);
                    }
                }
            }
            // Additional languages
            "cs" | "scala" | "pl" | "pm" => {
                if let Ok(source_code) = fs::read_to_string(file_path) {
                    if let Ok(extracted) = self.symbol_extractor.extract_symbols(&source_code, file_path) {
                        symbols.extend(extracted);
                    }
                }
            }
            _ => {
                // Unsupported file extension - just store filename/dirname symbols
            }
        }

        Ok(symbols)
    }

    pub fn get_symbols_by_file(&self, file_path: &Path) -> Vec<CodeSymbol> {
        self.symbols_cache.get(file_path).cloned().unwrap_or_default()
    }

    pub fn get_all_symbols(&self) -> Vec<CodeSymbol> {
        self.symbols_cache.values().flatten().cloned().collect()
    }

    pub async fn index_directory(&mut self, directory: &Path, patterns: &[String]) -> anyhow::Result<()> {
        // Collect all valid file paths first
        let mut file_paths = Vec::new();
        let walker = self.gitignore_filter.create_walker(directory);
        
        for entry in walker.build() {
            if let Some(path) = self.gitignore_filter.should_process_entry(&entry) {
                if self.file_filter.matches_patterns(&path, patterns) && self.file_filter.should_index_file(&path) {
                    file_paths.push(path);
                }
            }
        }
        
        // Process files in parallel using rayon
        let verbose = self.verbose;
        let results: Vec<_> = file_paths
            .par_iter()
            .filter_map(|path| {
                // Create a temporary indexer for parallel processing
                let mut temp_indexer = TreeSitterIndexer::with_options(verbose, self.respect_gitignore);
                if temp_indexer.initialize_sync().is_err() {
                    if verbose {
                        eprintln!("Warning: Failed to initialize indexer for {}", path.display());
                    }
                    return None;
                }
                
                match temp_indexer.index_file_sync(path) {
                    Ok(symbols) => Some((path.clone(), symbols)),
                    Err(e) => {
                        if verbose {
                            eprintln!("Warning: Failed to index {}: {}", path.display(), e);
                        }
                        None
                    }
                }
            })
            .collect();
        
        // Merge results back into main indexer
        for (path, symbols) in results {
            self.symbols_cache.insert(path, symbols);
        }
        
        Ok(())
    }

    pub fn clear_cache(&mut self) {
        self.symbols_cache.clear();
    }
}